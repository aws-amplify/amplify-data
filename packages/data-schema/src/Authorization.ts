import type { UnionToIntersection } from '@aws-amplify/data-schema-types';

const __data = Symbol('data');

/**
 * All possible providers.
 *
 * This list should not be used if you need to restrict available providers
 * according to an auth strategcy. E.g., `public` auth can only be facilitated
 * by `apiKey` and `iam` providers.
 */
export const Providers = [
  'apiKey',
  'iam',
  'userPools',
  'oidc',
  'function',
] as const;
export type Provider = (typeof Providers)[number];

/**
 * The subset of auth providers that can facilitate `public` auth.
 */
export const PublicProviders = ['apiKey', 'iam'] as const;
export type PublicProvider = (typeof PublicProviders)[number];

/**
 * The subset of auth providers that can facilitate `private` auth.
 */
export const PrivateProviders = ['userPools', 'oidc', 'iam'] as const;
export type PrivateProvider = (typeof PrivateProviders)[number];

/**
 * The subset of auth providers that can facilitate `owner` auth.
 */
export const OwnerProviders = ['userPools', 'oidc'] as const;
export type OwnerProviders = (typeof OwnerProviders)[number];

/**
 * The subset of auth providers that can facilitate `group` auth.
 */
export const GroupProviders = ['userPools', 'oidc'] as const;
export type GroupProvider = (typeof GroupProviders)[number];

/**
 * The subset of auth providers that can facilitate `custom` auth.
 */
export const CustomProviders = ['function'] as const;
export type CustomProvider = (typeof CustomProviders)[number];

export const Strategies = [
  'public',
  'private',
  'owner',
  'groups',
  'custom',
] as const;
export type Strategy = (typeof Strategies)[number];

/**
 * The operations that can be performed against an API.
 */
export const Operations = [
  'create',
  'update',
  'delete',
  'read',
  'get',
  'list',
  'sync',
  'listen',
  'search',
] as const;
export type Operation = (typeof Operations)[number];

export type Authorization<
  AuthStrategy extends Strategy,
  AuthField extends string | undefined,
  AuthFieldPlurality extends boolean,
> = {
  [__data]: {
    strategy?: AuthStrategy;
    provider?: Provider;
    operations?: Operation[];
    groupOrOwnerField?: AuthField;
    groups?: string[];
    multiOwner: AuthFieldPlurality;
    identityClaim?: string;
    groupClaim?: string;
  };
};

export type OwnerField = object;

type BuilderMethods<T extends object> = {
  [K in keyof T as T[K] extends (...args: any) => any ? K : never]: T[K];
};

/**
 * Creates a shallow copy of an object with an individual field pruned away.
 *
 * @param original The original object to prune.
 * @param without The field to prune.
 * @returns The pruned object.
 */
function omit<T extends object, O extends string>(
  original: T,
  without: O,
): Omit<T, O> {
  const pruned = { ...original };
  delete (pruned as any)[without];
  return pruned;
}

function to<SELF extends Authorization<any, any, any>>(
  this: SELF,
  operations: Operation[],
) {
  (this as any)[__data].operations = operations;
  return omit(this, 'to');
}

function inField<
  SELF extends Authorization<any, any, any>,
  Field extends string,
>(this: SELF, field: Field) {
  this[__data].groupOrOwnerField = field;
  const built = omit(this, 'inField');

  return built as unknown as BuilderMethods<typeof built> &
    Authorization<
      SELF[typeof __data]['strategy'],
      Field,
      SELF[typeof __data]['multiOwner']
    >;
}

/**
 * Specifies a property of the identity JWT to use in place of `sub::username`
 * as the value to match against the owner field for authorization.
 *
 * @param this Authorization object to operate against.
 * @param property A property of identity JWT.
 * @returns A copy of the Authorization object with the claim attached.
 */
function identityClaim<SELF extends Authorization<any, any, any>>(
  this: SELF,
  property: string,
) {
  this[__data].identityClaim = property;
  return omit(this, 'identityClaim');
}

function withClaimIn<SELF extends Authorization<any, any, any>>(
  this: SELF,
  property: string,
) {
  this[__data].groupClaim = property;
  return omit(this, 'withClaimIn');
}

function validateProvider(
  needle: Provider | undefined,
  haystack: readonly Provider[],
) {
  if (needle && !haystack.includes(needle)) {
    throw new Error(`Invalid provider (${needle}) given!`);
  }
}

function authData<
  Strat extends Strategy = 'public',
  Field extends string | undefined = undefined,
  isMulti extends boolean = false,
  Builders extends object = object,
>(
  defaults: Partial<Authorization<Strat, Field, isMulti>[typeof __data]>,
  builderMethods: Builders,
): Authorization<Strat, Field, isMulti> & Builders {
  return {
    [__data]: {
      strategy: 'public',
      provider: undefined,
      operations: undefined,
      groupOrOwnerField: undefined,
      multiOwner: false,
      identityClaim: undefined,
      groups: undefined,
      ...defaults,
    } as any,
    ...builderMethods,
  };
}

/**
 * Defines an authorization rule for your data models and fields. First choose an authorization strategy (`public`,
 * `private`, `owner`, `group`, or `custom`), then choose an auth provider (`apiKey`, `iam`, `userPools`, `oidc`, or `function`)
 * and optionally use `.to(...)` to specify the operations that can be performed against your data models and fields.
 */
export const allow = {
  /**
   * Authorize unauthenticated users. By default, `.public()` uses an API key based authorization. You can additionally
   * use `.public('iam')` to use IAM based authorization for unauthenticated users.
   * @param provider the authentication provider - supports "apiKey" or "iam" as valid providers
   * @returns an authorization rule for unauthenticated users
   */
  public(provider?: PublicProvider) {
    validateProvider(provider, PublicProviders);
    return authData(
      {
        strategy: 'public',
        provider,
      },
      {
        to,
      },
    );
  },

  /**
   * Authorize authenticated users. By default, `.private()` uses an Amazon Cognito user pool based authorization. You can additionally
   * use `.private("iam")` or `.private("oidc")` to use IAM or OIDC based authorization for authenticated users.
   * @param provider the authentication provider - supports "userPools", "iam", or "oidc"
   * @returns an authorization rule for authenticated users
   */
  private(provider?: PrivateProvider) {
    validateProvider(provider, PrivateProviders);
    return authData(
      {
        strategy: 'private',
        provider,
      },
      {
        to,
      },
    );
  },

  /**
   * Authorize access on a per-user (owner) basis. By setting owner-based authorization, a new `owner: a.string()`
   * field will be added to the model to store which user "owns" the item. Upon item creation, the "owner field" is
   * auto-populated with the authenticated user's information. You can which field should be used as the owner field
   * by chaining the * `.inField(...)` method.
   *
   * By default, `.owner()` uses an Amazon Cognito user pool based authorization. You can additionally
   * use `.owner("oidc")` to use OIDC based authentication to designate the owner.
   *
   * To change the specific claim that should be used as the user identifier within the owner field, chain the
   * `.identityClaim(...)` method.
   *
   * @param provider the authentication provider - supports "userPools", "iam", or "oidc"
   * @returns an authorization rule for authenticated users
   */
  owner(provider?: OwnerProviders) {
    validateProvider(provider, OwnerProviders);
    return authData(
      {
        strategy: 'owner',
        provider,
        groupOrOwnerField: 'owner',
      },
      {
        to,
        inField,
        identityClaim,
      },
    );
  },

  /**
   * Authorize access for multi-user / multi-owner access. By setting multi-owner-based authorization, a new `owners: a.string().array()`
   * field will be added to the model to store which users "own" the item. Upon item creation, the "owners field" is
   * auto-populated with the authenticated user's information. To grant other users access to the item, append their user identifier into the `owners` array.
   * You can which field should be used as the owners field by chaining the * `.inField(...)` method.
   *
   * By default, `.multipleOwners()` uses an Amazon Cognito user pool based authorization. You can additionally
   * use `.multipleOwners("oidc")` to use OIDC based authentication to designate the owner.
   *
   * To change the specific claim that should be used as the user identifier within the owners field, chain the
   * `.identityClaim(...)` method.
   *
   * @param provider the authentication provider - supports "userPools", "iam", or "oidc"
   * @returns an authorization rule for authenticated users
   */
  multipleOwners(provider?: OwnerProviders) {
    validateProvider(provider, OwnerProviders);
    return authData(
      {
        strategy: 'owner',
        multiOwner: true,
        provider,
        groupOrOwnerField: 'owner',
      },
      {
        to,
        inField,
        identityClaim,
      },
    );
  },

  /**
   * Authorize a specific user group. Provide the name of the specific user group to have access.
   *
   * By default, `.specificGroup()` uses an Amazon Cognito user pool based authorization. You can additionally
   * use `.specificGroup("group-name", "oidc")` to use OIDC based authentication to designate the user group.
   *
   * To change the specific claim that should be used as the user group identifier, chain the
   * `.withClaimIn(...)` method.
   * @param group the name of the group to authorize
   * @param provider the authentication provider - supports "userPools" or "oidc"
   * @returns an authorization rule to grant access by a specific group
   */
  specificGroup(group: string, provider?: GroupProvider) {
    return authData(
      {
        strategy: 'groups',
        provider,
        groups: [group],
      },
      {
        to,
        withClaimIn,
      },
    );
  },

  /**
   * Authorize multiple specific user groups. Provide the names of the specific user groups to have access.
   *
   * By default, `.specificGroups()` uses an Amazon Cognito user pool based authorization. You can additionally
   * use `.specificGroups(["group-a", "group-b"], "oidc")` to use OIDC based authentication to designate the user group.
   *
   * To change the specific claim that should be used as the user group identifier, chain the
   * `.withClaimIn(...)` method.
   * @param groups the names of the group to authorize defined as an array
   * @param provider the authentication provider - supports "userPools" or "oidc"
   * @returns an authorization rule to grant access by a specific group
   */
  specificGroups(groups: string[], provider?: GroupProvider) {
    return authData(
      {
        strategy: 'groups',
        provider,
        groups,
      },
      {
        to,
        withClaimIn,
      },
    );
  },

  /**
   * Authorize if a user is part of a group defined in a data model field.
   *
   * By default, `.groupDefinedIn()` uses an Amazon Cognito user pool based authorization. You can additionally
   * use `.groupDefinedIn("field-name", "oidc")` to use OIDC based authentication to designate the user group.
   *
   * To change the specific claim that should be used as the user group identifier within the groups field, chain the
   * `.withClaimIn(...)` method.
   * @param groupsField the field that should store the authorized user group information
   * @param provider the authentication provider - supports "userPools" or "oidc"
   * @returns an authorization rule to grant access by a specific group
   */
  groupDefinedIn<T extends string>(groupsField: T, provider?: GroupProvider) {
    return authData(
      {
        strategy: 'groups',
        provider,
        groupOrOwnerField: groupsField,
      },
      {
        to,
      },
    );
  },

  /**
   * Authorize if a user is part of a one of the groups defined in a data model field.
   *
   * By default, `.groupsDefinedIn()` uses an Amazon Cognito user pool based authorization. You can additionally
   * use `.groupsDefinedIn("field-name", "oidc")` to use OIDC based authentication to designate the user group.
   *
   * To change the specific claim that should be used as the user group identifier within the groups field, chain the
   * `.withClaimIn(...)` method.
   * @param groupsField the field that should store the list of authorized user groups
   * @param provider the authentication provider - supports "userPools" or "oidc"
   * @returns an authorization rule to grant access by a specific group
   */
  groupsDefinedIn<T extends string>(groupsField: T, provider?: GroupProvider) {
    return authData(
      {
        strategy: 'groups',
        provider,
        groupOrOwnerField: groupsField,
        multiOwner: true,
      },
      {
        to,
      },
    );
  },

  custom(provider?: CustomProvider) {
    return authData(
      {
        strategy: 'custom',
        provider,
      },
      {
        to,
      },
    );
  },
} as const;

/**
 * Turns the type from a list of `Authorization` rules like this:
 *
 * ```typescript
 * [
 *  allow.public(),
 *  allow.owner().inField('otherfield'),
 *  allow.multipleOwners().inField('editors')
 * ]
 * ```
 *
 * Into a union of the possible `fieldname: type` auth objects like this:
 *
 * ```typescript
 * {
 *  owner?: string | undefined;
 * } | {
 *  otherfield?: string | undefined;
 * } | {
 *  editors?: string[] | undefined;
 * }
 * ```
 */
export type ImpliedAuthField<T extends Authorization<any, any, any>> =
  T extends Authorization<infer _Strat, infer Field, infer isMulti>
    ? Field extends undefined
      ? never
      : Field extends string
      ? isMulti extends true
        ? { [K in Field]?: string[] }
        : { [K in Field]?: string }
      : never
    : never;

/**
 * Turns the type from a list of `Authorization` rules like this:
 *
 * ```typescript
 * [
 *  allow.public(),
 *  allow.owner().inField('otherfield'),
 *  allow.multipleOwners().inField('editors')
 * ]
 * ```
 *
 * Into an object type that includes all auth fields like this:
 *
 * ```typescript
 * {
 *  owner?: string | undefined;
 *  otherfield?: string | undefined;
 *  editors?: string[] | undefined;
 * }
 * ```
 */
export type ImpliedAuthFields<T extends Authorization<any, any, any>> =
  ImpliedAuthField<T> extends never
    ? never
    : UnionToIntersection<ImpliedAuthField<T>>;

export const accessData = <T extends Authorization<any, any, any>>(
  authorization: T,
) => authorization[__data];
