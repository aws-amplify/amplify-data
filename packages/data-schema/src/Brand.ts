const brand = Symbol('brand');
type brand = typeof brand;

export type Brand<T, TBrand extends string> = T & {
  [brand]: TBrand;
};

export function brandTarget<T extends object, TBrand extends string>(
  target: T,
  brandName: TBrand,
): Brand<T, TBrand> {
  return { ...target, [brand]: brandName };
}
