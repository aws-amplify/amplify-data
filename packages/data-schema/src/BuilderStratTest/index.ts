export {
  RefType as RefDataBranding,
  ref as refBranding,
  getData as getDataBranding,
} from './RefDataBranding';
export {
  RefType as RefDataKBranding,
  ref as refKBranding,
  getData as getDataKBranding,
} from './RefDataClassBranding';
export {
  RefType as RefInternalCasting,
  InternalRef as InternalRefInternalCasting,
  ref as refInternalCasting,
} from './RefInternalCasting';
