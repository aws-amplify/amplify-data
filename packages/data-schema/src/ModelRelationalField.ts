import { SetTypeSubArg } from '@aws-amplify/data-schema-types';
import { Brand } from './util';
import { Authorization } from './Authorization';

/**
 * Used to "attach" auth types to ModelField without exposing them on the builder.
 */
export const __auth = Symbol('__auth');

const brandName = 'modelRelationalField';

export enum ModelRelationshipTypes {
  hasOne = 'hasOne',
  hasMany = 'hasMany',
  belongsTo = 'belongsTo',
  manyToMany = 'manyToMany',
}

type RelationshipTypes = `${ModelRelationshipTypes}`;

const arrayTypeRelationships = ['hasMany', 'manyToMany'];

type ModelRelationalFieldData = {
  fieldType: 'model';
  type: ModelRelationshipTypes;
  relatedModel: string;
  array: boolean;
  valueRequired: boolean;
  arrayRequired: boolean;
  relationName?: string;
  references?: string[];
  authorization: Authorization<any, any, any>[];
};

export type ModelRelationalFieldParamShape = {
  type: 'model';
  relationshipType: string;
  relatedModel: string;
  array: boolean;
  valueRequired: boolean;
  references?: string[];
  arrayRequired: boolean;
  relationName?: string;
};

type ModelRelationalFieldFunctions<
  T extends ModelRelationalFieldParamShape,
  // RM adds structural separation with ModelField; easier to identify it when mapping to ClientTypes
  RM extends string | symbol,
  UsedMethod extends keyof ModelRelationalField<T, RM> = never,
> = {
  /**
   * When set, it requires the value of the relationship type to be required.
   */
  valueRequired(): ModelRelationalField<
    SetTypeSubArg<T, 'valueRequired', true>,
    UsedMethod | 'valueRequired'
  >;
  /**
   * Reference sets the foreign key on which to establish the relationship
   */
  references(
    references: string[],
  ): ModelRelationalField<
    SetTypeSubArg<T, 'references', string[]>,
    UsedMethod | 'references'
  >;
  /**
   * When set, it requires the relationship to always return a value
   */
  required(): ModelRelationalField<
    // The RM generic cannot be "required" since no such field exists
    SetTypeSubArg<T, 'arrayRequired', true>,
    UsedMethod | 'required'
  >;
  /**
   * When set, it requires the relationship to always return an array value
   * @deprecated this modifier should not be used and will be removed
   * in the next minor version of this package.
   */
  arrayRequired(): ModelRelationalField<
    SetTypeSubArg<T, 'arrayRequired', true>,
    UsedMethod | 'arrayRequired'
  >;
  /**
   * Configures field-level authorization rules. Pass in an array of authorizations `(a.allow.____)` to mix and match
   * multiple authorization rules for this field.
   */
  authorization<AuthRuleType extends Authorization<any, any, any>>(
    rules: AuthRuleType[],
  ): ModelRelationalField<
    T,
    UsedMethod | 'authorization',
    UsedMethod,
    AuthRuleType
  >;
};

export type ModelRelationalField<
  T extends ModelRelationalFieldParamShape,
  // RM adds structural separation with ModelField; easier to identify it when mapping to ClientTypes
  RM extends string | symbol,
  K extends keyof ModelRelationalField<T, RM> = never,
  Auth = undefined,
> = Omit<ModelRelationalFieldFunctions<T, RM, K>, K> & {
  // This is a lie. This property is never set at runtime. It's just used to smuggle auth types through.
  [__auth]?: Auth;
} & Brand<typeof brandName>;

/**
 * Internal representation of Model Field that exposes the `data` property.
 * Used at buildtime.
 */
export type InternalRelationalField = ModelRelationalField<
  ModelRelationalFieldParamShape,
  string,
  never
> & {
  data: ModelRelationalFieldData;
};

const relationalModifiers = [
  'required',
  'arrayRequired',
  'valueRequired',
  'references',
  'authorization',
] as const;

const relationModifierMap: Record<
  `${ModelRelationshipTypes}`,
  (typeof relationalModifiers)[number][]
> = {
  belongsTo: ['authorization', 'references'],
  hasMany: ['arrayRequired', 'valueRequired', 'authorization', 'references'],
  hasOne: ['required', 'authorization', 'references'],
  manyToMany: ['arrayRequired', 'valueRequired', 'authorization'],
};

export type RelationTypeFunctionOmitMapping<
  Type extends ModelRelationshipTypes,
> = Type extends ModelRelationshipTypes.belongsTo
  ? 'required' | 'arrayRequired' | 'valueRequired'
  : Type extends ModelRelationshipTypes.hasMany
    ? 'required'
    : Type extends ModelRelationshipTypes.hasOne
      ? 'arrayRequired' | 'valueRequired'
      : Type extends ModelRelationshipTypes.manyToMany
        ? 'required' | 'references'
        : never;

function _modelRelationalField<
  T extends ModelRelationalFieldParamShape,
  RelatedModel extends string,
  RT extends ModelRelationshipTypes,
>(type: RT, relatedModel: RelatedModel, relationName?: string) {
  const data: ModelRelationalFieldData = {
    relatedModel,
    type,
    fieldType: 'model',
    array: false,
    valueRequired: false,
    arrayRequired: false,
    references: undefined,
    relationName,
    authorization: [],
  };

  if (arrayTypeRelationships.includes(type)) {
    data.array = true;
  }
  const relationshipBuilderFunctions = {
    required() {
      data.arrayRequired = true;

      return this;
    },
    arrayRequired() {
      data.arrayRequired = true;

      return this;
    },
    valueRequired() {
      data.valueRequired = true;

      return this;
    },
    references(references: string[]) {
      data.references = references;

      return this;
    },
    authorization(rules) {
      data.authorization = rules;

      return this;
    },
  } as ModelRelationalField<T, RelatedModel>;

  const builder = Object.fromEntries(
    relationModifierMap[type].map((key) => [
      key,
      relationshipBuilderFunctions[key],
    ]),
  );

  return {
    ...builder,
    data,
  } as InternalRelationalField as ModelRelationalField<
    T,
    RelatedModel,
    RelationTypeFunctionOmitMapping<typeof type>
  >;
}

export type ModelRelationalTypeArgFactory<
  RM extends string,
  RT extends RelationshipTypes,
  IsArray extends boolean,
  RelationName extends string | undefined = undefined,
> = {
  type: 'model';
  relatedModel: RM;
  relationshipType: RT;
  array: IsArray;
  valueRequired: false;
  arrayRequired: false;
  relationName: RelationName;
  references: string[];
};

/**
 * Create a one-directional one-to-one relationship between two models using the `hasOne("MODEL_NAME")` method.
 * A hasOne relationship always uses a reference to the related model's identifier. Typically this is the `id` field
 * unless overwritten with the `identifier()` method.
 * @param relatedModel the name of the related model
 * @returns a one-to-one relationship definition
 */
export function hasOne<RM extends string>(relatedModel: RM) {
  return _modelRelationalField<
    ModelRelationalTypeArgFactory<RM, ModelRelationshipTypes.hasOne, false>,
    RM,
    ModelRelationshipTypes.hasOne
  >(ModelRelationshipTypes.hasOne, relatedModel);
}

/**
 * Create a one-directional one-to-many relationship between two models using the `hasMany()` method.
 * @param relatedModel the name of the related model
 * @returns a one-to-many relationship definition
 */
export function hasMany<RM extends string>(relatedModel: RM) {
  return _modelRelationalField<
    ModelRelationalTypeArgFactory<RM, ModelRelationshipTypes.hasMany, true>,
    RM,
    ModelRelationshipTypes.hasMany
  >(ModelRelationshipTypes.hasMany, relatedModel);
}

/**
 * Make a `hasOne()` or `hasMany()` relationship bi-directional using the `belongsTo()` method.
 * The belongsTo() method requires that a hasOne() or hasMany() relationship already exists from
 * parent to the related model.
 * @param relatedModel name of the related `.hasOne()` or `.hasMany()` model
 * @returns a belong-to relationship definition
 */
export function belongsTo<RM extends string>(relatedModel: RM) {
  return _modelRelationalField<
    ModelRelationalTypeArgFactory<RM, ModelRelationshipTypes.belongsTo, false>,
    RM,
    ModelRelationshipTypes.belongsTo
  >(ModelRelationshipTypes.belongsTo, relatedModel);
}

/**
 * Create a many-to-many relationship between two models with the manyToMany() method.
 * Provide a common relationName on both models to join them into a many-to-many relationship.
 * Under the hood a many-to-many relationship is modeled with a "join table" with corresponding
 * `hasMany()` relationships between the two related models. You must set the same `manyToMany()`
 * field on both models of the relationship.
 * @param relatedModel name of the related model
 * @param opts pass in the `relationName` that will serve as the join table name for this many-to-many relationship
 * @returns a many-to-many relationship definition
 */
export function manyToMany<RM extends string, RN extends string>(
  relatedModel: RM,
  opts: { relationName: RN },
) {
  return _modelRelationalField<
    ModelRelationalTypeArgFactory<
      RM,
      ModelRelationshipTypes.manyToMany,
      true,
      RN
    >,
    RM,
    ModelRelationshipTypes.manyToMany
  >(ModelRelationshipTypes.manyToMany, relatedModel, opts.relationName);
}
