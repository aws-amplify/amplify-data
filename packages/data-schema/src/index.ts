import { schema } from './ModelSchema';
import { model } from './ModelType';
import {
  id,
  string,
  integer,
  float,
  boolean,
  date,
  time,
  datetime,
  timestamp,
  email,
  json,
  phone,
  url,
  ipAddress,
} from './ModelField';
import { ref } from './RefType';
import { hasOne, hasMany, belongsTo, manyToMany } from './ModelRelationalField';
import { allow } from './Authorization';
import { customType } from './CustomType';
import { enumType } from './EnumType';
import { query, mutation, subscription } from './CustomOperation';
import { handler } from './Handler';

// We are re-exporting with this pattern to make the JSDoc happy.
// Otherwise the JSDoc wouldn't consistently show up as a customer types
// "a."
export {
  schema,
  model,
  ref,
  customType,
  enumType as enum,
  query,
  mutation,
  subscription,
  hasOne,
  hasMany,
  belongsTo,
  manyToMany,
  allow,
  id,
  string,
  integer,
  float,
  boolean,
  date,
  time,
  datetime,
  timestamp,
  email,
  json,
  phone,
  url,
  ipAddress,
  handler,
};
