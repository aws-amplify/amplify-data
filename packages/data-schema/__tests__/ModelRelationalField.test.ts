import { a } from '../index';
import { InternalField } from '../src/ModelField';
import { InternalRelationalField } from '../src/ModelRelationalField';

describe('relational field required modifier', () => {
  describe('belongsTo', () => {
    it("doesn't offer a required modifier", () => {
      expect(() => {
        // @ts-expect-error
        const field = a.belongsTo('Test').required();
      }).toThrow();
    });

    it("doesn't offer an arrayRequired modifier", () => {
      expect(() => {
        // @ts-expect-error
        const field = a.belongsTo('Test').arrayRequired();
      }).toThrow();
    });

    it("doesn't offer an valueRequired modifier", () => {
      expect(() => {
        // @ts-expect-error
        const field = a.belongsTo('Test').valueRequired();
      }).toThrow();
    });

    it('offers an authorization modifier', () => {
      expect(() => {
        const field = a.belongsTo('Test').authorization([]);
      }).not.toThrow();
    });
  });

  describe('hasOne', () => {
    it('offers a required modifier', () => {
      expect(() => {
        const field = a.hasOne('Test').required();
      }).not.toThrow();
    });

    it("doesn't offer an arrayRequired modifier", () => {
      expect(() => {
        // @ts-expect-error
        const field = a.hasOne('Test').arrayRequired();
      }).toThrow();
    });

    it("doesn't offer an valueRequired modifier", () => {
      expect(() => {
        // @ts-expect-error
        const field = a.hasOne('Test').valueRequired();
      }).toThrow();
    });

    it('offers an authorization modifier', () => {
      expect(() => {
        const field = a.hasOne('Test').authorization([]);
      }).not.toThrow();
    });

    it('required sets arrayRequired to true', () => {
      const field = a.hasOne('Test').required() as InternalRelationalField;
      expect(field.data.arrayRequired).toBe(true);
    });
  });

  describe('hasMany', () => {
    it("doesn't offer a required modifier", () => {
      expect(() => {
        // @ts-expect-error
        const field = a.hasMany('Test').required();
      }).toThrow();
    });

    it('offers an arrayRequired modifier', () => {
      expect(() => {
        const field = a.hasMany('Test').arrayRequired();
      }).not.toThrow();
    });

    it('offers an valueRequired modifier', () => {
      expect(() => {
        const field = a.hasMany('Test').valueRequired();
      }).not.toThrow();
    });

    it('offers an authorization modifier', () => {
      expect(() => {
        const field = a.hasMany('Test').authorization([]);
      }).not.toThrow();
    });

    it('arrayRequired sets arrayRequired to true', () => {
      const field = a
        .hasMany('Test')
        .arrayRequired() as InternalRelationalField;
      expect(field.data.arrayRequired).toBe(true);
    });

    it('valueRequired sets valueRequired to true', () => {
      const field = a
        .hasMany('Test')
        .valueRequired() as InternalRelationalField;
      expect(field.data.valueRequired).toBe(true);
    });
  });

  describe('manyToMany', () => {
    it("doesn't offer a required modifier", () => {
      expect(() => {
        // @ts-ignore
        const field = a.manyToMany('Test', { relationName: 'Test' }).required();
      }).toThrow();
    });

    it('offers an arrayRequired modifier', () => {
      expect(() => {
        const field = a
          .manyToMany('Test', { relationName: 'Test' })
          .arrayRequired();
      }).not.toThrow();
    });

    it('offers an valueRequired modifier', () => {
      expect(() => {
        const field = a
          .manyToMany('Test', { relationName: 'Test' })
          .valueRequired();
      }).not.toThrow();
    });

    it('offers an authorization modifier', () => {
      expect(() => {
        const field = a
          .manyToMany('Test', { relationName: 'Test' })
          .authorization([]);
      }).not.toThrow();
    });

    it('arrayRequired sets arrayRequired to true', () => {
      const field = a
        .manyToMany('Test', { relationName: 'Test' })
        .arrayRequired() as InternalRelationalField;
      expect(field.data.arrayRequired).toBe(true);
    });

    it('valueRequired sets valueRequired to true', () => {
      const field = a
        .manyToMany('Test', { relationName: 'Test' })
        .valueRequired() as InternalRelationalField;
      expect(field.data.valueRequired).toBe(true);
    });
  });
});
