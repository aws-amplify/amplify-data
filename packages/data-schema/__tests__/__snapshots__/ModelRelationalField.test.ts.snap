// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`schema generation with relationships ddb hasOne / belongsTo explicitly defined reference field on related model is supported 1`] = `
"type Team @model @auth(rules: [{allow: public}])
{
  motto: String
  project: Project @hasOne(references: ["teamId"])
}

type Project @model @auth(rules: [{allow: public}])
{
  name: String
  teamId: ID
  team: Team @belongsTo(references: ["teamId"])
}"
`;

exports[`schema generation with relationships ddb masMany / belongsTo explicitly defined reference field on related model is supported 1`] = `
"type Team @model @auth(rules: [{allow: public}])
{
  motto: String
  members: [Member] @hasMany(references: ["teamId"])
}

type Member @model @auth(rules: [{allow: public}])
{
  name: String
  teamId: ID
  team: Team @belongsTo(references: ["teamId"])
}"
`;

exports[`schema generation with relationships ddb masMany / belongsTo fails if reference field on related model is undefined 1`] = `
"type Team @model @auth(rules: [{allow: public}])
{
  motto: String
  members: [Member] @hasMany(references: ["teamId"])
}

type Member @model @auth(rules: [{allow: public}])
{
  name: String
  team: Team @belongsTo(references: ["teamId"])
}"
`;

exports[`schema generation with relationships ddb masMany / belongsTo partition key + sort key - undefined sort key reference on related model fails 1`] = `
"type Team @model @auth(rules: [{allow: public}])
{
  id: ID! @primaryKey(sortKeyFields: ["sk"])
  sk: ID!
  motto: String
  members: [Member] @hasMany(references: ["teamId","teamSk"])
}

type Member @model @auth(rules: [{allow: public}])
{
  name: String
  teamId: ID
  team: Team @belongsTo(references: ["teamId","teamSk"])
}"
`;

exports[`schema generation with relationships ddb masMany / belongsTo partition key + sort key is supported 1`] = `
"type Team @model @auth(rules: [{allow: public}])
{
  id: ID! @primaryKey(sortKeyFields: ["sk"])
  sk: ID!
  motto: String
  members: [Member] @hasMany(references: ["teamId","teamSk"])
}

type Member @model @auth(rules: [{allow: public}])
{
  name: String
  teamId: ID
  teamSk: ID
  team: Team @belongsTo(references: ["teamId","teamSk"])
}"
`;

exports[`schema generation with relationships relationships 1`] = `
"type Team @model @auth(rules: [{allow: public}])
{
  motto: String
  members: [Member] @hasMany(references: ["teamId"])
  project: Project @hasOne(references: ["teamId"])
}

type Project @model @auth(rules: [{allow: public}])
{
  name: String
  teamId: ID
  team: Team @belongsTo(references: ["teamId"])
}

type Member @model @auth(rules: [{allow: public}])
{
  name: String
  teamId: ID
  team: Team @belongsTo(references: ["teamId"])
}"
`;

exports[`schema generation with relationships sql references 1`] = `
"type Team @model @auth(rules: [{allow: public}])
{
  id: ID! @primaryKey
  motto: String
  members: [Member] @hasMany(references: ["teamMembersId"])
  project: Project @hasOne(references: ["teamProjectId"])
}

type Project @model @auth(rules: [{allow: public}])
{
  id: ID! @primaryKey
  name: String
  teamId: ID
  team: Team @belongsTo(references: ["teamId"])
}

type Member @model @auth(rules: [{allow: public}])
{
  id: ID! @primaryKey
  name: String
  teamId: ID
  team: Team @belongsTo(references: ["teamId"])
}"
`;
